airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/S3ConsumerFactory.kt

```diff
@@ -140,6 +121,46 @@ class S3ConsumerFactory {
         }
     }
 
+    fun createAsync(
+        outputRecordCollector: Consumer<AirbyteMessage>,
+        storageOps: S3StorageOperations,
+        s3Config: S3DestinationConfig,
+        catalog: ConfiguredAirbyteCatalog
+    ): SerializedAirbyteMessageConsumer? {
+        val writeConfigs = createWriteConfigs(storageOps, s3Config, catalog)
+        // Buffer creation function: yields a file buffer that converts
+        // incoming data to the correct format for the destination.
+        val createFunction =
+            getCreateFunction(
+                s3Config,
+                Function<String, BufferStorage> { fileExtension: String ->
+                    FileBuffer(fileExtension)
+                }
+            )
+        return AsyncStreamConsumer(
+            outputRecordCollector,
+            onStartFunction(storageOps, writeConfigs),
+            onCloseFunction(storageOps, writeConfigs),
+            S3DestinationFlushFunction(
+                // Ensure the file buffer is always larger than the memory buffer,
+                // as the file buffer will be flushed at the end of the memory flush.
+                optimalBatchSizeBytes = (FileBuffer.MAX_PER_STREAM_BUFFER_SIZE_BYTES * 0.9).toLong()
+            ) {
+                // Yield a new BufferingStrategy every time we flush (for thread-safety).
+                SerializedBufferingStrategy(
```

Comment:
tbh I don't think I fully grok the implications here. AFAICT this seems strictly better? since every flush requires us to open a new output stream anyway?

Reply:
`SerializedBufferingStrategy` maintains one file buffer per stream. So, in theory, each flush worker could write to it and it could maintain separate buffers, which could be configured to flush at the optimal time. 

But even if we force a file buffer flush at the end of call to the flush function (current behavior), we would still be sharing the same object across threads, which means the whole class and all its subclasses, file buffers, etc would have to be audited for thread safety. (The naive implementation crashes immediately because of concurrent attempts to access a the hash map that stores the buffers.)

Reply:
aha. that makes sense. So it's an improvement over what we currently do, but it makes it harder to do better things in the future 👍